src/Servers/Kestrel/Core/src/Internal/Http/Http1ContentLengthMessageBody.cs

```diff
@@ -213,17 +214,23 @@ public override void AdvanceTo(SequencePosition consumed, SequencePosition exami
 
             _isReading = false;
 
+            // The current body is read, though there might be more bytes to read on the stream with pipelining
             if (_readCompleted)
             {
-                // If the old stored _readResult was canceled, it's already been observed. Do not store a canceled read result permanently.
-                _readResult = new ReadResult(_readResult.Buffer.Slice(consumed, _readResult.Buffer.End), isCanceled: false, isCompleted: true);
-
-                if (!_finalAdvanceCalled && _readResult.Buffer.Length == 0)
+                if (!_finalAdvanceCalled && consumed.Equals(_readResult.Buffer.End))
```

Comment:
I should have caught this in the PR to `main`, but I don't think we can compare `SequencePosition`s in this way. The problem is that `consumed` could theoretically point to the very end of a `ReadOnlySequenceSegment<byte>` and `_readResult.Buffer.End` might point to the very beginning of the next `ReadOnlySequenceSegment<byte>`. All the bytes would be consumed in this case, but the `SequencePosition`s wouldn't be equal.

I think we need to preemptively `_readResult.Buffer.Slice(consumed, _readResult.Buffer.End)` but not use the resulting buffer if it `IsEmpty`.

Reply:
Yes, use the length 👍🏾